File: PHP/lib/QueryType.php
<?php

require_once 'lib/nvps.php';
/**
 * Description of QueryType
 *
 * @author Roy Keynan
 */
abstract class QueryType {

    function __construct() {
    }

    abstract protected function getQueryType();

    abstract protected function addQueryParametersToTheList($nvps);

    public function addParametersToTheNvps($nvps){
        $nvps->add("queryType", $this->getQueryType());
        $this->addQueryParametersToTheList($nvps);
    }

    protected function addDatesToNvps($nvps , $startDate , $endDate){
        if ($startDate != NULL){
            $nvps->add("startDate", $startDate);
            if($endDate != NULL){
                $nvps->add("endDate", $endDate);
            }
        }
    }
}

File: Python/zooz/QueryType.py
__author__ = 'Roy Keynan'

import abc
from abc import ABCMeta


def add_dates_to_plist(plist, start_date, end_date):
    if start_date is not None:
        plist["startDate"] = start_date
        if end_date is not None:
            plist["endDate"] = end_date


class QueryType(object):
    __metaclass__ = ABCMeta

    @abc.abstractmethod
    def __init__(self):
        pass

    @abc.abstractmethod
    def get_query_type(self):
        pass

    @abc.abstractmethod
    def add_query_parameters_to_the_list(self, plist):
        pass

    def add_parameters_to_the_list(self, plist):
        plist["queryType"] = self.get_query_type()
        self.add_query_parameters_to_the_list(plist)

